pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -1029,111 +1029,57 @@ func combineTaskRunAndTaskSpecAnnotations(pr *v1beta1.PipelineRun, pipelineTask
+	// This works because all finally tasks run in parallel, so there is no need to consider time spent by other finally tasks
+	if finallyTimeout == nil || finallyTimeout.Duration == apisconfig.NoTimeoutDuration {
+		return taskRunTimeout
 	}
-
-	// If the value of the timeout is 0 for any resource, there is no timeout.
-	// It is impossible for pr.Spec.Timeout to be nil, since SetDefault always assigns it with a value.
-	taskRunTimeout = taskRunTimeoutHelper(timeout, pr, taskRunTimeout, rprt)
-
-	// Now that we know if we still have time to run the final task, subtract tasksTimeout if needed
-	if taskRunTimeout.Duration > time.Second {
-		taskRunTimeout.Duration -= tasksTimeout
+	if finallyTimeout.Duration < taskRunTimeout.Duration {
+		return finallyTimeout
 	}
-
 	return taskRunTimeout
 }
 
+// getTaskRunTimeout returns the timeout to set when creating the ResolvedPipelineRunTask.
+// If there is no timeout for the TaskRun, returns 0.
+// If pipeline level timeouts have already been exceeded, returns 1 second.
 func getTaskRunTimeout(ctx context.Context, pr *v1beta1.PipelineRun, rprt *resources.ResolvedPipelineRunTask) *metav1.Duration {
-	var taskRunTimeout = &metav1.Duration{Duration: apisconfig.NoTimeoutDuration}
-
 	var timeout time.Duration
-
-	switch {
-	case pr.Spec.Timeout != nil:
-		timeout = pr.Spec.Timeout.Duration
-	case pr.Spec.Timeouts != nil:
-		if pr.Spec.Timeouts.Tasks != nil {
-			timeout = pr.Spec.Timeouts.Tasks.Duration
-			break
-		}
-
-		if pr.Spec.Timeouts.Pipeline != nil {
-			timeout = pr.Spec.Timeouts.Pipeline.Duration
-		}
-
-		if pr.Spec.Timeouts.Finally != nil {
-			timeout -= pr.Spec.Timeouts.Finally.Duration
-		}
-	default:
-		defaultTimeout := time.Duration(config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes)
-		timeout = defaultTimeout * time.Minute
+	if pr.TasksTimeout() != nil {
+		timeout = pr.TasksTimeout().Duration
+	} else {
+		timeout = pr.GetTimeout(ctx)
 	}
-
-	// If the value of the timeout is 0 for any resource, there is no timeout.
-	// It is impossible for pr.Spec.Timeout to be nil, since SetDefault always assigns it with a value.
-	taskRunTimeout = taskRunTimeoutHelper(timeout, pr, taskRunTimeout, rprt)
-
-	return taskRunTimeout
+	return taskRunTimeoutHelper(timeout, pr, rprt)
 }
 
-func taskRunTimeoutHelper(timeout time.Duration, pr *v1beta1.PipelineRun, taskRunTimeout *metav1.Duration, rprt *resources.ResolvedPipelineRunTask) *metav1.Duration {
+// taskRunTimeoutHelper returns the timeout to set when creating the ResolvedPipelineRunTask.
+// `timeout` is the pipelinerun Tasks timeout (for a Task) or the pipelinerun timeout (for a finally Task, or if the Tasks timeout is not set).
+// If there is no timeout for the TaskRun, returns 0.
+// If pipeline level timeouts have already been exceeded, returns 1 second.
+func taskRunTimeoutHelper(timeout time.Duration, pr *v1beta1.PipelineRun, rprt *resources.ResolvedPipelineRunTask) *metav1.Duration {
```

Comment:
I realize that `taskRunTimeoutHelper` was the name of the function from before, but I suggest that we rename it to something more descriptive as part of this refactor (maybe `calculateTaskRunTimeout`)